TYPO3.Flow/Classes/TYPO3/Flow/Configuration/ConfigurationManager.php

```diff
@@ -651,24 +651,51 @@ protected function postProcessConfiguration(array &$configurations)
         foreach ($configurations as $key => $configuration) {
             if (is_array($configuration)) {
                 $this->postProcessConfiguration($configurations[$key]);
-            } elseif (is_string($configuration)) {
-                $matches = array();
-                preg_match_all('/(?:%)((?:\\\?[\d\w_\\\]+\:\:)?[A-Z_0-9]+)(?:%)/', $configuration, $matches);
-                if (count($matches[1]) > 0) {
-                    foreach ($matches[1] as $match) {
-                        if (defined($match)) {
-                            if ($configurations[$key] === '%' . $match . '%') {
-                                // the constant expression spans the complete directive, assign directly to keep type
-                                $configurations[$key] = constant($match);
-                            } else {
-                                // the constant is only a substring of the directive, replace that part accordingly
-                                $configurations[$key] = str_replace('%' . $match . '%', constant($match), $configurations[$key]);
-                            }
-                        }
-                    }
-                }
             }
+            if (!is_string($configuration)) {
+                continue;
+            }
+            $configurations[$key] = $this->replaceVariablesInValue($configuration);
+        }
+    }
+
+    /**
+     * Replaces variables (in the format %CONSTANT% or %ENV::ENVIRONMENT_VARIABLE)
+     * in the given (configuration) value string.
+     *
+     * @param string $value
+     * @return mixed
+     */
+    protected function replaceVariablesInValue($value)
+    {
+        $matches = [];
+        $replacements = [];
+        $replacementCount = 0;
+
+        preg_match_all('/(?:%)((?:\\\?[\d\w_\\\]+\:\:)?[A-Z_0-9]+)(?:%)/', $value, $matches);
+        array_walk($matches[1], function ($match, $key) use ($matches, &$replacements, &$replacementCount) {
+            $replacementCount++;
+
+            if (defined($match)) {
+                $replacements[$matches[0][$key]] = constant($match);
```

Comment:
Had a little bit hard time to find what ``$matches[0][$key]`` actually is. Maybe we could assign that to an expressive variable beforehand, sth. like "$matchWithPercentages = ..."